server/image_pull.go

```diff
@@ -61,6 +64,14 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 		var tmpImg types.ImageCloser
 		tmpImg, err = s.StorageImageServer().PrepareImage(&sourceCtx, img)
 		if err != nil {
+			// We're not able to find the image remotely, check if it's
+			// available locally, but only for localhost/ prefixed ones
+			if strings.HasPrefix(img, localRegistryPrefix) {
+				if _, err := s.StorageImageServer().GetStore().Image(img); err == nil {
```

Comment:
Could this use `s.StorageImageServer().ImageStatus()` instead? It’s a bit more expensive, but it has clearly the same semantics, unlike reaching through to the storage directly (I’m primarily worried about the extra lookups `storageReference.resolveImage` does, although the layering violation is not great either).

`ImageStatus()` is more expensive than `GetStore().Image()`, OTOH it is called after exiting the loop anyway, so if efficiency were a concern, the two can be deduplicated (both here and in the “already in store, skipping pull” path below), making `ImageStatus` ultimately cheaper.

Reply:
Implemented as suggested and I also added a comment regarding the image pull policy.